sdk/core/System.ClientModel/src/Pipeline/ClientRetryPolicy.cs

```diff
@@ -0,0 +1,238 @@
+            }
+
+            bool shouldRetry = async ?
+                await ShouldRetryAsync(message, thisTryException).ConfigureAwait(false) :
+                ShouldRetry(message, thisTryException);
+
+            if (shouldRetry)
+            {
+                TimeSpan delay = GetNextDelay(message, message.RetryCount);
+                if (delay > TimeSpan.Zero)
+                {
+                    if (async)
+                    {
+                        await WaitAsync(delay, message.CancellationToken).ConfigureAwait(false);
+                    }
+                    else
+                    {
+                        Wait(delay, message.CancellationToken);
+                    }
+                }
+
+                // Dispose the content stream to free up a connection if the request has any
+                message.Response?.ContentStream?.Dispose();
+
+                message.RetryCount++;
+                OnTryComplete(message);
+
+                continue;
+            }
+
+            if (thisTryException != null)
+            {
+                // Rethrow if there's only one exception.
+                if (allTryExceptions!.Count == 1)
+                {
+                    ExceptionDispatchInfo.Capture(thisTryException).Throw();
+                }
+
+                throw new AggregateException($"Retry failed after {message.RetryCount + 1} tries.", allTryExceptions);
+            }
+
+            // ShouldRetry returned false this iteration and the last request
+            // we sent didn't cause an exception to be thrown.
+            // So, we're done.  Exit the while loop.
+            break;
+        }
+    }
+
+    protected virtual void OnSendingRequest(PipelineMessage message) { }
+
+    protected virtual ValueTask OnSendingRequestAsync(PipelineMessage message) => default;
+
+    protected virtual void OnRequestSent(PipelineMessage message) { }
+
+    protected virtual ValueTask OnRequestSentAsync(PipelineMessage message) => default;
+
+    protected virtual void OnTryComplete(PipelineMessage message) { }
+
+    internal bool ShouldRetry(PipelineMessage message, Exception? exception)
+    {
+        // If there was no exception and we got a success response, don't retry.
+        if (exception is null && message.Response is not null && !message.Response.IsError)
+        {
+            return false;
+        }
+
+        return ShouldRetryCore(message, exception);
+    }
+
+    internal async ValueTask<bool> ShouldRetryAsync(PipelineMessage message, Exception? exception)
+    {
+        // If there was no exception and we got a success response, don't retry.
+        if (exception is null && message.Response is not null && !message.Response.IsError)
+        {
+            return false;
+        }
+
+        return await ShouldRetryCoreAsync(message, exception).ConfigureAwait(false);
+    }
+
+    protected virtual bool ShouldRetryCore(PipelineMessage message, Exception? exception)
```

Comment:
Why does this need to be suffixed with Core?

Reply:
Yeah, I see the problem here, and that you flagged here: https://github.com/Azure/azure-sdk-for-net/pull/41325#discussion_r1457690403

I don't know if we want to expose the ShouldRetry method publicly, but for Azure.Core back-compat, I do think we want an extensibility point, and I think we'd rather use the template method for this kind of extensibility.  Tell you what - I'll make ShouldRetry and GetNextDelay both public so that I'm correctly following the [Template Method Pattern](https://learning.oreilly.com/library/view/framework-design-guidelines/9780135896457/ch09.xhtml#sec9_9), and we can have the conversation with architects regarding whether this is too many APIs to expose on the ClientRetryPolicy type.  It'll be interesting to look at from that perspective one way or another -- these are the implications of not having a separate type to encapsulate the delay logic.